src/win32/win32_keychain.c

```diff
@@ -0,0 +1,280 @@
+           name, type & NCRYPT_IMPL_HARDWARE_FLAG ? "with" : "without");
+}
+
+static void init() {
+    static uv_once_t once_init;
+    uv_once(&once_init, do_init);
+}
+
+static wchar_t* name_to_wchar(const char *name) {
+    int wlen = MultiByteToWideChar(CP_UTF8, 0, name, (int) strlen(name), NULL, 0);
+    wchar_t *wname = calloc(wlen, sizeof(wchar_t) + 1);
+    MultiByteToWideChar(CP_UTF8, 0, name, (int)strlen(name), wname, wlen);
+    return wname;
+}
+
+static int gen_key(keychain_key_t *key, enum keychain_key_type type, const char *name){
+    init();
+    static wchar_t *algos[] = {
+            NCRYPT_ECDSA_P521_ALGORITHM,
+            NCRYPT_ECDSA_P384_ALGORITHM,
+            NCRYPT_ECDSA_P256_ALGORITHM,
+            NCRYPT_ECDSA_ALGORITHM,
+    };
+
+    if (type != keychain_key_ec) {
+        UM_LOG(ERR, "unsupported key type: %d", type);
+        return -1;
+    }
+
+    wchar_t *wname = name_to_wchar(name);
+    int res = -1;
+
+    NCRYPT_KEY_HANDLE h = 0;
+    wchar_t *algo = NULL;
+    for (int i = 0; i < sizeof(algos)/sizeof(algos[0]); i++) {
+        if (NCryptIsAlgSupported(provider, algos[i], 0) == ERROR_SUCCESS) {
+            algo = algos[i];
+            break;
+        }
+    }
+    UM_LOG(DEBG, "generating key with algo[%ls]", algo);
+    SECURITY_STATUS rc = NCryptCreatePersistedKey(provider, &h, algo, wname, 0, 0);
+
+    if (rc != ERROR_SUCCESS) {
+        UM_LOG(ERR, "failed to generate key[%ls]: %s", wname, win32_error(rc));
+        goto done;
+    }
+    rc = NCryptFinalizeKey(h, 0);
+    if (rc != ERROR_SUCCESS) {
+        UM_LOG(ERR, "failed to generate key[%ls]: %s", wname, win32_error(rc));
+        goto done;
+    }
+
+    *key = (keychain_key_t) h;
+    res = 0;
+
+    done:
+    free(wname);
+    return res;
+}
+
+static int load_key(keychain_key_t *k, const char *name) {
+    init();
+    wchar_t *wname = name_to_wchar(name);
+
+    NCRYPT_KEY_HANDLE keyh = 0;
+    SECURITY_STATUS rc = NCryptOpenKey(provider, &keyh, wname, 0, 0);
+
+    free(wname);
+    
+    if (rc == ERROR_SUCCESS) {
+        *k = (keychain_key_t) keyh;
+        return 0;
+    }
+    
+    UM_LOG(ERR, "failed to load key[%s]: %s", name, win32_error(rc));
+
+    return -1;
+}
+
+static void free_key(keychain_key_t key){
+    NCRYPT_KEY_HANDLE h = (NCRYPT_KEY_HANDLE) key;
+    if (NCryptIsKeyHandle(h)) {
+        NCryptFreeObject(h);
+    }
+}
+
+static int rem_key(const char *name){
+    init();
+
+    wchar_t *wname = name_to_wchar(name);
+    int res = 0;
+
+    NCRYPT_KEY_HANDLE keyh = 0;
+    SECURITY_STATUS rc = NCryptOpenKey(provider, &keyh, wname, 0, 0);
+
+    if (rc == ERROR_SUCCESS) {
+        rc = NCryptDeleteKey(keyh, 0);
+        if (rc != ERROR_SUCCESS) {
+            UM_LOG(WARN, "failed to delete key[%s]: %s", name, win32_error(rc));
+            res = -1;
+        }
+    } else {
+        UM_LOG(WARN, "failed to open key[%s]: %s", name, win32_error(rc));
```

Comment:
failed to open key still should return 0?

Reply:
it is removing key, so it's ok if key is not found